File: ios/PurchasesHybridCommon/PurchasesHybridCommon/NSDate+HybridAdditions.m
//
// Created by RevenueCat.
// Copyright (c) 2020 Purchases. All rights reserved.
//

#import "NSDate+HybridAdditions.h"

NS_ASSUME_NONNULL_BEGIN

static NSDateFormatter *formatter;
static dispatch_once_t onceToken;

static NSString *stringFromDate(NSDate *date) {
    dispatch_once(&onceToken, ^{
        // Here we're not using NSISO8601DateFormatter as we need to support iOS < 10
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        dateFormatter.timeZone = [NSTimeZone timeZoneWithAbbreviation:@"GMT"];
        dateFormatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss'Z'";
        dateFormatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
        formatter = dateFormatter;
    });

    return [formatter stringFromDate:date];
}

@implementation NSDate (RCExtensions)

- (NSString *)rc_formattedAsISO8601 {
    return stringFromDate(self);
}

- (double)rc_millisecondsSince1970AsDouble {
    return [self timeIntervalSince1970] * 1000.0;
}

@end


NS_ASSUME_NONNULL_END

File: ios/PurchasesHybridCommon/PurchasesHybridCommon/NSDate+HybridAdditions.swift
//
//  NSDate+HybridAdditions.swift
//  PurchasesHybridCommon
//
//  Created by Andrés Boedo on 4/13/22.
//  Copyright © 2022 RevenueCat. All rights reserved.
//

import Foundation

// note: This extension is only temporary, since we're
// dropping iOS 11 in the upcoming major
extension Date {

    func rc_formattedAsISO8601() -> String {
        return Self.stringFromDate(self)
    }

    func rc_millisecondsSince1970AsDouble() -> Double {
        return self.timeIntervalSince1970 * 1000.0
    }

}

private extension Date {

    static func stringFromDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.timeZone = NSTimeZone(abbreviation: "GMT") as TimeZone?
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter.string(from: date as Date)
    }

}
